src/nodes/display/ToonOutlineNode.js

```diff
@@ -0,0 +1,264 @@
+import TempNode from '../core/TempNode.js';
+import { float, nodeObject, normalize, vec4 } from '../tsl/TSLBase.js';
+import { NodeUpdateType } from '../core/constants.js';
+import { Color } from '../../math/Color.js';
+import NodeMaterial from '../../materials/nodes/NodeMaterial.js';
+import { cameraProjectionMatrix } from '../../nodes/accessors/Camera.js';
+import { modelViewMatrix } from '../../nodes/accessors/ModelNode.js';
+import { positionLocal } from '../../nodes/accessors/Position.js';
+import { normalLocal } from '../../nodes/accessors/Normal.js';
+import { BackSide } from '../../constants.js';
+
+class ToonOutlineNode extends TempNode {
+
+	static get type() {
+
+		return 'ToonOutlineNode';
+
+	}
+
+	constructor( scenePassNode, colorNode, thicknessNode, alphaNode ) {
+
+		super( 'vec4' );
+
+		this.scenePassNode = scenePassNode;
+		this.colorNode = colorNode;
+		this.thicknessNode = thicknessNode;
+		this.alphaNode = alphaNode;
+
+		//
+
+		this._camera = this.scenePassNode.camera;
+		this._scene = this.scenePassNode.scene;
+
+		this._traverseCallbackPrepare = this._prepare.bind( this );
+		this._traverseCallbackRestore = this._restore.bind( this );
+
+		this._originalMaterials = new Map();
+		this._materialCache = new WeakMap();
+		this._visibilityCache = new WeakMap();
+
+
+		//
+
+		this.updateBeforeType = NodeUpdateType.FRAME;
+
+	}
+
+	dispose() {
+
+		this._scene = null;
+
+		this._originalMaterials.clear();
+		this._materialCache = new WeakMap();
+		this._visibilityCache = new WeakMap();
+
+	}
+
+	updateBefore( frame ) {
+
+		const { renderer } = frame;
+		const camera = this._camera;
+		const scene = this._scene;
+
+		const currentAutoClear = renderer.autoClear;
+		const currentSceneAutoUpdate = scene.matrixWorldAutoUpdate;
+		const currentSceneBackground = scene.background;
+		const currentShadowMapEnabled = renderer.shadowMap.enabled;
+		const currentRenderTarget = renderer.getRenderTarget();
+
+		scene.matrixWorldAutoUpdate = false;
+		scene.background = null;
+		renderer.autoClear = false;
+		renderer.shadowMap.enabled = false;
+
+		scene.traverse( this._traverseCallbackPrepare );
+
+		renderer.setRenderTarget( this.scenePassNode.renderTarget );
+		renderer.render( scene, camera );
```

Comment:
Instead of rendering, I tried a `renderer.clear()` to delete the contents of `scenePassNode.renderTarget` for testing. But unfortunately this operation has no effect. I wonder if the render target is correctly bound or a different reference error causes issue.

Reply:
Could we use `renderer.setRenderObjectFunction()` to render the outline and then render the target object?

```js
renderer.setRenderObjectFunction( ( object, scene, camera, geometry, material, group, lightsNode ) => {

	// outline
	const outlineMaterial = ...
	renderer.renderObject( object, scene, camera, geometry, outlineMaterial, group, lightsNode );

	// target
	renderer.renderObject( object, scene, camera, geometry, material, group, lightsNode );

} );
```

Reply:
If `ToonOutlineNode` would make use of `renderer.setRenderObjectFunction()`, then the target objects are rendered more than once because they are already rendered in the normal scene pass. 

So `ToonOutlineNode` should just render the outlines. If `ToonOutlineNode` shares the same render target (and thus depth buffer), this should normally work as expected like with `WebGLRenderer`.

Reply:
I think in this case the `ToonOutlineNode` class should be extended from `PassNode`. The problem with `traverse()` is that it ignores the frustrum and makes the code more complex.

Reply:
Deriving from `PassNode` and using `setRenderObjectFunction()` greatly simplifies the code! I love the new approach 😊 .